alibi_detect/utils/loading.py

```diff
@@ -0,0 +1,625 @@
+    """
+    # Before main resolution, update filepaths relative to config file
+    if config_dir is not None:
+        for key in DIR_FIELDS:
+            src = _get_nested_value(cfg, key)
+            if isinstance(src, str):
+                src = config_dir.joinpath(Path(src))
+                if src.is_file() or src.is_dir():
+                    _set_nested_value(cfg, key, str(src))
+
+    # Resolve filepaths (load files) and resolve function/object registries
+    for key in FIELDS_TO_RESOLVE:
+        logger.info('Resolving config field: {}.'.format(key))
+
+        src = _get_nested_value(cfg, key)
+
+        obj = None
+        # Resolve runtime registered function/object
+        if isinstance(src, str):
+            if src.startswith('@'):
+                src = src[1:]
+                if src in registry.get_all():
+                    obj = registry.get(src)
+                else:
+                    raise ValueError("Can't find %s in the custom function registry" % src)
+                logger.info('Successfully resolved registry entry %s' % src)
+
+            # Load dill or numpy file
+            elif Path(src).is_file():
+                if Path(src).suffix == '.dill':
+                    obj = dill.load(open(src, 'rb'))
+                if Path(src).suffix == '.npy':
+                    obj = np.load(src)
+
+#            # Pytorch device  # TODO
+#            elif key[-1] == 'device':
+#                obj = set_device(src)
+
+        # Resolve dict spec
+        elif isinstance(src, dict):
+            backend = cfg.get('backend', 'tensorflow')
+            if key[-1] in ('model', 'proj'):
+                obj = _load_model_config(src, backend)
+            if key[-1] == 'embedding':
+                obj = _load_embedding(src, backend)
+            elif key[-1] == 'tokenizer':
+                obj = _load_tokenizer(src)
+            elif key[-1] == 'optimizer':
+                obj = _load_optimizer(src, backend)
+
+        # Put the resolved function into the cfg dict
+        if obj is not None:
+            _set_nested_value(cfg, key, obj)
+
+    # Convert selected lists to tuples
+    for key in FIELDS_TO_TUPLE:
+        val = _get_nested_value(cfg, key)
+        if val is not None:
+            _set_nested_value(cfg, key, tuple(val))
+
+    # Convert selected lists to np.ndarray's
+    for key in FIELDS_TO_ARRAY:
+        val = _get_nested_value(cfg, key)
+        if val is not None:
+            _set_nested_value(cfg, key, np.array(val))
+
+    # Convert selected str's to required dtype's
+    for key in FIELDS_TO_DTYPE:
+        val = _get_nested_value(cfg, key)
+        if val is not None:
+            val = val.split('.')
+#            val[0] = np if val[0] == 'np' else tf if val[0] == 'tf' else torch if val[0] == 'torch' else None
+            # TODO - add above back in once optional deps are handled properly
+            val[0] = np if val[0] == 'np' else tf if val[0] == 'tf' else None
```

Comment:
something like:
```py
lib, dtype, *_ = val.split('.')
lib = np if lib == 'np' else tf if lib == 'tf' else None
_set_nested_value(cfg, key, getattr(lib, dtype))
```
would be more readable here.

Reply:
Thanks, much nicer!